[RFC]Memkind integration prefix solution

michalbiesek opened the pull request:
Second approach to problem with distinguish pmem or ram, first is realized in PR #85
The distinguish iis on Redis side with saving information about destination during allocation

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pmem/redis/86)
<!-- Reviewable:end -->

zhouyuan committed:
[RFC][DNM] Allow to store data on PMEM

Sample code for hooks on string/list. The general idea is to move
object content to DCPMM with small changes.

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

zhouyuan committed:
allow to support more structure

Signed-off-by: Yuan Zhou <yuan.zhou@intel.com>

Commit:
Distinguish allocation between ram and pmem using prefix

Commit:
Update pmem configuration

- create Pmem kind on start
- destroy pmem kind just before end
- provide configuration parameter of pmem

Commit:
Update allocator structure regarding current state of repo

zhouyuan commented on src/zmalloc.c:
```diff
@@ -170,6 +198,22 @@ void *zrealloc(void *ptr, size_t size) {
+#ifdef USE_MEMKIND
+    void* new_ptr = NULL;
+    if (ptr) {
+        new_ptr = zrealloc_local((void*)((char*)(ptr) - MEMKIND_PREFIX_SIZE), size + MEMKIND_PREFIX_SIZE);
```
Is it possible to make use of the same logic in zfree here?  
```
uint64_t *is_ram = (uint64_t*)((char*)(ptr) - MEMKIND_PREFIX_SIZE);
if (*is_ram) {
  zrealloc_local()
} else {
  mrealloc()
}
```

michalbiesek replied:
Please see newest commit (https://github.com/pmem/redis/pull/86/commits/5256136524c2ca1e336e6e4c62d2cfb11d1106c0)

Commit:
Update zmalloc structure with realloc function

- realloc function use the same logic as free - detect the place of
previous allocation ( pmem or RAM ) and calls properly function
- in case of calling realloc with NULL pointer it would behave like
malloc for RAM
